core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java

```diff
@@ -327,6 +334,79 @@ public synchronized void losePrimacy() {
     LOG.info("Raft server successfully restarted");
   }
 
+  @Override
+  public synchronized void checkpoint() throws IOException {
+    try {
+      mSnapshotAllowed.set(true);
+      CopycatClient client = createAndConnectClient();
+      long start = System.currentTimeMillis();
+      LOG.info("Submitting empty journal entry to trigger snapshot");
+      // New snapshot requires new segments (segment size is controlled by
+      // {@link PropertyKey#MASTER_JOURNAL_LOG_SIZE_BYTES_MAX}).
+      // If snapshot requirements are fulfilled, a snapshot will be triggered
+      // after sending an empty journal entry to Copycat
+      CompletableFuture<Void> future = client.submit(new JournalEntryCommand(
+          JournalEntry.getDefaultInstance()));
+      try {
+        future.get(1, TimeUnit.MINUTES);
+      } catch (TimeoutException | ExecutionException e) {
+        LOG.warn("Exception submitting entry to trigger snapshot: {}", e.toString());
+        throw new IOException("Exception submitting entry to trigger snapshot", e);
+      } catch (InterruptedException e) {
+        LOG.warn("Interrupted when submitting entry to trigger snapshot: {}", e.toString());
+        Thread.currentThread().interrupt();
+        throw new CancelledException("Interrupted when submitting entry to trigger snapshot", e);
+      }
+      waitForSnapshotStart(mStateMachine, start);
+      if (mStateMachine.isSnapshotting()) {
+        waitForSnapshotting(mStateMachine);
+      }
+    } finally {
+      mSnapshotAllowed.set(false);
+    }
+  }
+
+  /**
+   * Waits for snapshotting to start.
+   *
+   * @param stateMachine the journal state machine
+   * @param start the start time to check
+   */
+  private void waitForSnapshotStart(JournalStateMachine stateMachine,
+      long start) throws IOException {
+    try {
+      CommonUtils.waitFor("snapshotting to start", () ->
+              stateMachine.getLastSnapshotStartTime() > start,
+          WaitForOptions.defaults().setTimeoutMs(10 * Constants.SECOND_MS));
+      return;
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new CancelledException("Interrupted when waiting for snapshotting to start", e);
+    } catch (TimeoutException e) {
+      throw new IOException("Do not fulfill Copycat snapshot requirements. "
```

Comment:
this message will likely be propagated to users, who might not know about snapshots or copycat. Can we say something more general like "A recent checkpoint already exists, not creating a new one"?

Reply:
Yeah, changed. Now is more user-friendly.